kudu-handler/src/java/org/apache/hadoop/hive/kudu/KuduPredicateHandler.java

```diff
@@ -0,0 +1,175 @@
+    ExprNodeDesc residualPredicate = analyzer.analyzePredicate(predicateExpr, sConditions);
+
+    // Nothing to decompose.
+    if (sConditions.size() == 0) {
+      return null;
+    }
+
+    DecomposedPredicate decomposedPredicate = new DecomposedPredicate();
+    decomposedPredicate.pushedPredicate = analyzer.translateSearchConditions(sConditions);
+    decomposedPredicate.residualPredicate = (ExprNodeGenericFuncDesc) residualPredicate;
+    return decomposedPredicate;
+  }
+
+  /**
+   * Returns the list of Kudu predicates from the passed configuration.
+   *
+   * @param conf the execution configuration
+   * @param schema the schema of the Kudu table
+   * @return the list of Kudu predicates
+   */
+  public static List<KuduPredicate> getPredicates(Configuration conf, Schema schema) {
+    List<KuduPredicate> predicates = new ArrayList<>();
+    for (IndexSearchCondition sc : getSearchConditions(conf, schema)) {
+      predicates.add(conditionToPredicate(sc, schema));
+    }
+    return predicates;
+  }
+
+  private static List<IndexSearchCondition> getSearchConditions(Configuration conf, Schema schema) {
+    List<IndexSearchCondition> conditions = new ArrayList<>();
+    ExprNodeDesc filterExpr = getExpression(conf);
+    if (null == filterExpr) {
+      return conditions;
+    }
+    IndexPredicateAnalyzer analyzer = newAnalyzer(schema);
+    ExprNodeDesc residual = analyzer.analyzePredicate(filterExpr, conditions);
+
+    // There should be no residual since we already negotiated that earlier in
+    // decomposePredicate. However, with hive.optimize.index.filter
+    // OpProcFactory#pushFilterToStorageHandler pushes the original filter back down again.
+    // Since pushed-down filters are not omitted at the higher levels (and thus the
+    // contract of negotiation is ignored anyway), just ignore the residuals.
+    // Re-assess this when negotiation is honored and the duplicate evaluation is removed.
+    if (residual != null) {
+      LOG.debug("Ignoring residual predicate " + residual.getExprString());
+    }
+
+    return conditions;
+  }
+
+  private static ExprNodeDesc getExpression(Configuration conf) {
+    String filteredExprSerialized = conf.get(TableScanDesc.FILTER_EXPR_CONF_STR);
+    if (filteredExprSerialized == null) {
+      return null;
+    }
+    return SerializationUtilities.deserializeExpression(filteredExprSerialized);
+  }
+
+  private static KuduPredicate conditionToPredicate(IndexSearchCondition condition, Schema schema) {
+    ColumnSchema column = schema.getColumn(condition.getColumnDesc().getColumn());
+    GenericUDF genericUDF = condition.getOriginalExpr().getGenericUDF();
+    Object value = condition.getConstantDesc().getValue();
+
+    // Convert special Hive types.
+    if (value instanceof HiveDecimal) {
+      value = ((HiveDecimal) value).bigDecimalValue();
+    } else if (value instanceof Timestamp) {
+      value = ((Timestamp) value).toSqlTimestamp();
```

Comment:
I am not sure whether Kudu uses `local datetime` or `instant` for timestamp semantics (Hive uses `local datetime` in timestamp and `instant` in timestamp with local time zone, see https://cwiki.apache.org/confluence/display/Hive/Different+TIMESTAMP+types).
Note that this method just creates a `java.sql.Timestamp` with millis since epoch utc, i.e., calling `toString` method may not get you the correct String representation depending on the system timezone (you need to use a formatter).
Hive tests are run in PST. To verify that filtering on those values is working properly, insert timestamp values in `kudu_queries.q` that have less than 8 hours difference between them (in current tests, one value is in `2011` and another one in `2012`, which would not expose this issue even if it exists).

Reply:
Kudu currently has a type called "UNIXTIME_MICROS" that it maps timestamps to. I will update the tests to make sure I validate the handling.